### PR TITLE
Fix scaffolding unmapped type with underlying

### LINKED ISSUE #12016: RevEng: Using type alias of unmapped type on column causes NRE
See test in #11983

### PR DESCRIPTION
**Pull request by roji**
When scaffolding, a database column with an underlying store type but which itself has no mapping caused an NRE.

Fixes #12016

### CODE REVIEW

File: src/EFCore.Design/Scaffolding/Internal/RelationalScaffoldingModelFactory.cs [hunk 1]
```diff
@@ -856,6 +856,11 @@ protected virtual TypeScaffoldingInfo GetTypeScaffoldingInfo([NotNull] DatabaseC
                 column.IsKeyOrIndex(),
                 column.IsRowVersion());
 
+            if (typeScaffoldingInfo != null)
```
Comment (bricelam) [inline src/EFCore.Design/Scaffolding/Internal/RelationalScaffoldingModelFactory.cs]: `==`?

**Comment by roji** [comment]
@bricelam, pushed a corrected fix with a test.

However, I spent an hour trying to make the tests run (both on release/2.1 and on dev) and couldn't, either in Rider or on the command-line (dotnet test as well as build.sh). The Jenkins build also seems to be broken for unrelated reasons. Unfortunately I don't have time to dive into this at the moment, so I haven't actually tested - hopefully you can take a look and correct if needed.

Thanks and sorry...

**Comment by bricelam** [comment]
Thanks! I'll dig into the failures

**Comment by bricelam** [comment]
@roji Is this something we should consider patching in 2.1.1?

**Comment by roji** [comment]
Thanks.

I'm not sure what how high your bar is for patches... this is really an edge case that I doubt people will actually run into. On the other hand it seems extremely low-risk...

File: src/EFCore.Design/Scaffolding/Internal/RelationalScaffoldingModelFactory.cs [hunk 2]
```diff
@@ -856,6 +856,11 @@ protected virtual TypeScaffoldingInfo GetTypeScaffoldingInfo([NotNull] DatabaseC
                 column.IsKeyOrIndex(),
                 column.IsRowVersion());
 
+            if (typeScaffoldingInfo == null)
+            {
+                return null;
+            }
+
             if (column.GetUnderlyingStoreType() != null)
```
[the remark this answers is not in the document]
Comment (roji) [reply src/EFCore.Design/Scaffolding/Internal/RelationalScaffoldingModelFactory.cs]: You're right @smitpatel (as I said I couldn't actually run the test, sorry).

Regarding the nit, I just code this way in case any more logic is added later. The moment that `typeScaffoldingInfo` we know that we're not going to be able to scaffold.

**Comment by roji** [comment]
In case you're interested, let me give some context on how I ran into this...

PostgreSQL has enum types. While version 2.1 of the Npgsql provider supports mapping CLR enums to PG enums, at least at the moment it isn't possible to scaffold them - fully scaffolding database enums would involve scaffolding a C# enum type as well, which is theoretically a nice idea but is a bit out of scope of the current scaffolding. As a result, when NpgsqlDatabaseModelFactory detects an enum, it logs a warning and skips it - enum columns are *not* included in the resulting DatabaseModel. I did this because I wanted an informative log message specific for the enum case, rather then including the enum columns in the DatabaseModel and then have the scaffolding process log a more generic error message.

Now, a user recently complained (https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/issues/365) that trying to scaffold a database with an enum *type alias* triggers an NRE (type aliases are called domains in PostgreSQL). My enum detection code in NpgsqlDatabaseModelFactory did not identify enum domains, and so the DatabaseModel returned a column with a domain type (unknown type), with the underlying type being an enum (also unknown).

In other words, I think this bug is mainly (only?) a result of an issue in the provider's DatabaseModelFactory, which probably makes it non-critical for a 2.1.1 patch.

**Comment by smitpatel** [comment]
> In other words, I think this bug is mainly (only?) a result of an issue in the provider's DatabaseModelFactory, which probably makes it non-critical for a 2.1.1 patch.

Actually it is not bug in provider code. A type-aliased column whose underlying type is not mapped will cause above error. But it is slightly rare scenario so may not meet bar for patch. Filed #12016

**Comment by roji** [comment]
You're right.

**Comment by bricelam** [comment]
(Updated the test slightly since `bit` is a T-SQL type.)
